src/common/io.ts

```diff
@@ -154,38 +154,50 @@ export class ARIOReadable implements AoARIORead {
     });
   }
 
-  async getEpochSettings(params?: EpochInput): Promise<AoEpochSettings> {
-    const allTags = [
-      { name: 'Action', value: 'Epoch-Settings' },
-      {
-        name: 'Timestamp',
-        value:
-          (params as { timestamp?: number })?.timestamp?.toString() ??
-          Date.now().toString(),
-      },
-      {
-        name: 'Epoch-Index',
-        value: (params as { epochIndex?: number })?.epochIndex?.toString(),
-      },
-    ];
+  private async computeEpochIndexForTimestamp(
+    timestamp: number,
+  ): Promise<number> {
+    const epochSettings = await this.getEpochSettings();
+    const epochZeroStartTimestamp = epochSettings.epochZeroStartTimestamp;
+    const epochLengthMs = epochSettings.durationMs;
+    return Math.floor((timestamp - epochZeroStartTimestamp) / epochLengthMs);
+  }
 
-    return this.process.read<AoEpochSettings>({
+  private async computeEpochIndex(
+    params?: EpochInput,
+  ): Promise<string | undefined> {
+    const epochIndex = (params as { epochIndex?: number })?.epochIndex;
+    if (epochIndex !== undefined) {
+      return epochIndex.toString();
+    }
+
+    const timestamp = (params as { timestamp?: number })?.timestamp;
+    if (timestamp !== undefined) {
+      return (await this.computeEpochIndexForTimestamp(timestamp)).toString();
+    }
+
+    return undefined;
+  }
+
+  private epochSettings: AoEpochSettings | undefined;
```

Comment:
nit, put this at the top of the class with other attributes - it gets lost down here